### PR TITLE
reduce config duplication, update third-party actions

### LINKED ISSUE #100: Build args in dockerfiles should not use realistic values
## Description

Dockerfiles in this repo currently set realistic values as defaults for build arguments

https://github.com/rapidsai/dask-build-environment/blob/e121aa88f172c7cfe83c2774125fd9d33a145dc6/dask/Dockerfile#L1-L3

But never intentionally use those defaults... it's always expected that those build arguments will be fulfilled by values passed through to `docker build`:

https://github.com/rapidsai/dask-build-environment/blob/e121aa88f172c7cfe83c2774125fd9d33a145dc6/ci/gpuci/run.sh#L64

Those realistic values should be replaced with the literal string `unset` or similar.

## Benefits of this work

Reduces the risk of silently building and tagging an image that's using the wrong configuration (because of mistakes in passing through build args).

## Acceptance Criteria

* all configuration for build arguments is driver from outside of `Dockerfile`s

## Approach

This: https://github.com/rapidsai/docker/pull/689#discussion_r1684918803

## Notes

Created this issue based on https://github.com/rapidsai/dask-build-environment/pull/97#discussion_r1729432517

### PR DESCRIPTION
**Pull request by jameslamb**
Closes #100 

Follow-up to #97 

Contributes to https://github.com/rapidsai/build-planning/issues/40

Proposes the following changes:

* ~adding Python 3.12 images~
  - *~(now that those have been added upstream in https://github.com/rapidsai/ci-imgs/pull/137)~*
* removing as many sources of duplicated configuration as possible
  - *(with the goal of only needing to update `ci/axis/dask.yaml` in the future for changes like "support a new Python version")*
* updating all third-party GitHub Actions to their latest versions
  - *(which should resolve all the warnings you can see in job summaries, e.g. https://github.com/rapidsai/dask-build-environment/actions/runs/10293483059)
* adds a minimal `.gitignore` and `README.md`

### How do I test this?

I can't figure out where these image builds actually happen. At https://github.com/rapidsai/dask-build-environment/actions, I only see the "Check for gpuCI Updates" job running.

Could someone please tell me where `docker build` based on these `Dockerfiles` is run and how I could test these changes?

### CODE REVIEW

**Comment by jakirkham** [comment]
Thanks James! 🙏

Generally this seems reasonable to me. Asking Charles to take a closer look as he is the expert here 😉

**Comment by jameslamb** [comment]
I just reverted the Python 3.12 part of this. Realized that some of these images are actually installing RAPIDS packages, and we don't yet have packages for Python 3.12.

e.g.: https://github.com/rapidsai/dask-build-environment/blob/e121aa88f172c7cfe83c2774125fd9d33a145dc6/dask/environment.yml#L1-L14

The work to add Python 3.12 here is tracked in https://github.com/rapidsai/build-planning/issues/40

**Comment by charlesbluca** [comment]
Thanks @jameslamb! A lot of long-overdue improvements here.

> Could someone please tell me where docker build based on these Dockerfiles is run and how I could test these changes?

Yeah this is definitely something that could use more visibility - this is [echoed](https://github.com/rapidsai/dask-build-environment/blob/e121aa88f172c7cfe83c2774125fd9d33a145dc6/ci/gpuci/run.sh#L63-L64) as part of the build process, but this is only visible in the Jenkins [runs](https://gpuci.gpuopenanalytics.com/job/dask/job/dask-build-environment/job/branch/job/dask-build-env-main/BUILD_NAME=dask,CUDA_VER=11.8.0,LINUX_VER=ubuntu20.04,PYTHON_VER=3.10,RAPIDS_VER=24.10/1247/console), which I think are only visible to `@rapidsai/dask-build-environment-write`. Looking at one of these logs, the build command is:

```
docker build --pull -t gpuci/dask:24.10-cuda11.8.0-devel-ubuntu20.04-py3.10 --squash --build-arg RAPIDS_VER=24.10 --build-arg UCX_PY_VER=0.40 --build-arg CUDA_VER=11.8.0 --build-arg LINUX_VER=ubuntu20.04 --build-arg PYTHON_VER=3.10 /jenkins/workspace/dask/dask-build-environment/branch/dask-build-env-main/BUILD_NAME/dask/CUDA_VER/11.8.0/LINUX_VER/ubuntu20.04/PYTHON_VER/3.10/RAPIDS_VER/24.10/dask
```

Modifying this for local testing on this work (picking a tag arbitrarily):

```
docker build \
    --pull \
    -t gpuci/$PROJECT_NAME:$USER-dev \
    --squash \
    --build-arg RAPIDS_VER=24.10 \
    --build-arg UCX_PY_VER=0.40 \
    --build-arg CUDA_VER=11.8.0 \
    --build-arg LINUX_VER=ubuntu20.04 \
    --build-arg PYTHON_VER=3.10 \
    ./$PROJECT_NAME
```

**Comment by jameslamb** [comment]
Ah got it, thanks for that!

Yep, can confirm that I'm not able to view those Jenkins logs.

1. Could you tell me if it looks like these changes are passing CI?
2. If they are, could you approve and merge this?
3. Would you support folks from RAPIDS build-infra team being added with write access here, so we can help with these sorts of things like modifying the set of supported CUDA / Python verisons?

**Comment by charlesbluca** [comment]
> Could you tell me if it looks like these changes are passing CI?

Unfortunately, there isn't a PR builder job, so checking that these changes pass CI can't be done until this is merged in 😕 and typically I've relied on local testing for more extensive changes.

Locally, I can verify that not specifying all of the expected build args causes a failure:

```
docker build \
    --pull \
    -t gpuci/dask:charlesb-dev \
    ./dask
...
ERROR: failed to solve: rapidsai/miniforge-cuda:cudaunset-base-unset-pyunset: failed to resolve source metadata for docker.io/rapidsai/miniforge-cuda:cudaunset-base-unset-pyunset: docker.io/rapidsai/miniforge-cuda:cudaunset-base-unset-pyunset: not found
```

Then confirmed for each project that including those build arguments allowed the builds to proceed and pass.

I noticed that we missed this explicit setting of `RAPIDS_VER` in the [dask-image dockerfile](https://github.com/jameslamb/dask-build-environment/blob/73e2bafe0a39a467003f998228a8e4c361d507be/dask_image/Dockerfile#L10), which I think we can actually remove altogether since we don't actually install RAPIDS in that image - this would look like:

- dropping the setting of `RAPIDS_VER` from [this line](https://github.com/jameslamb/dask-build-environment/blob/73e2bafe0a39a467003f998228a8e4c361d507be/ci/gpuci/run.sh#L40) setting the dask-image docker build args
- dropping the [line](https://github.com/jameslamb/dask-build-environment/blob/73e2bafe0a39a467003f998228a8e4c361d507be/dask_image/Dockerfile#L10) I linked to above

Would you mind applying this change?

**Comment by charlesbluca** [comment]
> Would you support folks from RAPIDS build-infra team being added with write access here, so we can help with these sorts of things like modifying the set of supported CUDA / Python verisons?

Followed up with an internal request to do this for now, if that spins off into a relevant GitHub issue can follow up here

**Comment by jameslamb** [comment]
> *Unfortunately, there isn't a PR builder job, so checking that these changes pass CI can't be done until this is merged in 😕 and typically I've relied on local testing*

Ohhh ok, thanks for that. I thought that maybe I just couldn't see the PR builds because I didn't have sufficient permissions.

Thanks for testing this one locally.

> *think we can actually remove altogether since we don't actually install RAPIDS in that image*

Oh cool yeah, looks like that would be safe to do. I dropped `RAPIDS_VER` from `dask_image` in https://github.com/rapidsai/dask-build-environment/pull/101/commits/ddfd1e721cb6dbf3ee9bffff60975fc0b7a36a5c

File: .github/workflows/update-gpuci.yml
Comment (charlesbluca) [inline .github/workflows/update-gpuci.yml:1]: Noting that there isn't a super good way of testing this workflow prior to merging - was going to kick off a `workflow_dispatch` after merging to check that this still runs as expected with the latest versions, though maybe there's a better way to test a limited subset of this action (excluding actually opening a PR) on PRs that modify this file?

Okay with pushing that work off to a separate PR to avoid blocking the other changes here

**Review by charlesbluca — APPROVED** [review]
Thanks @jameslamb!

File: .github/workflows/update-gpuci.yml
Comment (jameslamb) [reply .github/workflows/update-gpuci.yml:1]: I think a manual run of just waiting for the next scheduled one (which will happen about 4 hours from now, at UTC midnight) is probably sufficient for testing changes like this. It seems to me that reverting changes like this would be pretty simple, and that the impact of this job being broken for on the order of a day is low.

Comment (charlesbluca) [reply .github/workflows/update-gpuci.yml:1]: Sounds good to me - in that case, this should be good to merge 👍🏼

**Pull request merged**
